[zoneminder] Added image channel, changed to jetty, Optimized performance

Mr-Eskildsen opened the pull request:
Signed-off-by: Martin Eskildsen <martin@eskildsen.name> (github:
Mr-Eskildsen)

ZoneMinder Binding
[zoneminder] Added Image channel to monitor
[zoneminder] Changed http client to jetty
[zoneminder] Optimized performance 



<!-- TESTING -->
JAR is available in the Eclipse IoT Marketplace, as well as here: 
http://www.eskildsen.name/openHAB/zoneminder/org.openhab.binding.zoneminder-2.3.0-SNAPSHOT.jar

martinvw commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/handler/ZoneMinderBaseThingHandler.java:
```diff
@@ -316,16 +361,24 @@ public Channel getChannel(ChannelUID channelUID) {
     public boolean isThingRefreshed() {
-        return thingRefreshed;
+        return (thingRefresh.get() > 0) ? false : true;
+        // return (thingRefresh > 0) ? false : true;
```
And check for any other commented out lines of code

martinvw commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/handler/ZoneMinderServerBridgeHandler.java:
```diff
@@ -50,69 +53,87 @@
+
             } catch (Exception exception) {
-                logger.error("[WATCHDOG]: Server run(): Exception: {}", exception.getMessage());
+                StackTraceElement ste = exception.getStackTrace()[0];
```
What are you doing here, do you really need to inspect the stacktrace, does it happen often or only incidental?

martinvw commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/handler/ZoneMinderServerBridgeHandler.java:
```diff
@@ -123,211 +144,173 @@ public void run() {
+
+                RefreshPriority cyclePriority = RefreshPriority.PRIORITY_ALARM;
+
+                // boolean isBatch = ((refreshCycleCount % intervalBatch) == 0);
```
Remove

martinvw commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/handler/ZoneMinderServerBridgeHandler.java:
```diff
@@ -123,211 +144,173 @@ public void run() {
-    }
+            initialized = true;
+            return;
+        } catch (Exception ex) {
```
Do you really need to catch the Exception-class or are there more specific exceptions you could catch like the IOException or maybe only RuntimeException's (which is not that good either). The current code and also catching RuntimeException's might 'hide' programming errors like NullPointerExceptions occurring.

https://www.google.nl/search?q=java%20why%20not%20to%20catch%20exception

Mr-Eskildsen replied:
Seems to be there onbly for historic reasons. Will be removed :-)

martinvw commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/handler/ZoneMinderServerBridgeHandler.java:
```diff
@@ -123,211 +144,173 @@ public void run() {
+            setConnected(false);
+            zoneMinderConnection = null;
+            setConnectionStatus(ZoneMinderConnectionStatus.UNINITIALIZED);
+        } catch (IllegalArgumentException | GeneralSecurityException | IOException | ZoneMinderUrlNotFoundException e) {
```
Can this small block of code really throw all of this on us?

Mr-Eskildsen replied:
Unfortunately yes. The problem is that the ZoneMinder API seems to always return a HTTP Code 401 , so I have to guess what ever the correct reason is. There is also a of combinations that can cause problems. 
* Authentication can be disabled, but API disabled => return code 401 , 
* Authentication can be enabled, wrong credentials  => return code 401 , 
* Authentication can be enabled, insufficent user rights in ZM => return code 401 , 
Just to mention a few, also things seems to differ in different ZoneMinder versions. I can't really see if authentication is enabled or disabled, just have to try and see what happends...

SInce the Jetty Client also uses exceptions, I did continue on that  road. To identify the correct combination I simply use trial and error. It might be possible to rewrite code to minimize usage of different exceptions. But it really requires intensive testing. 
This would be quite a change, so I would really avoid this for now.

martinvw commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/handler/ZoneMinderServerBridgeHandler.java:
```diff
@@ -336,33 +319,21 @@ protected String getThingId() {
-
-    protected ArrayList<IZoneMinderMonitorData> getMonitors(IZoneMinderSession session) {
-
+    protected ArrayList<IMonitorDataGeneral> getMonitors(IZoneMinderConnectionHandler session)
```
Make this return Type a List instead of an ArrayList

martinvw commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/handler/ZoneMinderServerBridgeHandler.java:
```diff
@@ -394,111 +365,140 @@ public void handleCommand(ChannelUID channelUID, Command command) {
     }
 
-    protected synchronized void refreshThing(IZoneMinderSession session, boolean fetchDiskUsage) {
+    protected /* synchronized */ void refreshThing(RefreshPriority refresh) {
```
remove commented out code

martinvw commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/handler/ZoneMinderServerBridgeHandler.java:
```diff
@@ -394,111 +365,140 @@ public void handleCommand(ChannelUID channelUID, Command command) {
+        RefreshPriority curPriority = RefreshPriority.DISABLED;
 
-        logger.debug("{}: 'refreshThing()': Thing='{}'!", getLogIdentifier(), this.getThing().getUID());
+        // logger.debug("{}: 'refreshThing()': Thing='{}'!", getLogIdentifier(), this.getThing().getUID());
```
remove commented out code

martinvw commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/handler/ZoneMinderServerBridgeHandler.java:
```diff
@@ -511,63 +511,50 @@ public synchronized Boolean isConnected() {
      * Set connection status.
      *
      * @param connected
+     * @throws ZoneMinderUrlNotFoundException
```
If you add this, please also mention a little bit about when the caller can expect them, and imho runtime exception do n not have to be mentioned there

martinvw commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/handler/ZoneMinderServerBridgeHandler.java:
```diff
@@ -618,276 +606,692 @@ private void onDisconnected() {
 
-    @Override
-    public void updateAvaliabilityStatus(IZoneMinderConnectionInfo connection) {
+    private int initRetriesCount = 0;
```
Please make sure that all the fields are one top of the class, or split up in different classes which have a higher cohesion

martinvw commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/handler/ZoneMinderServerBridgeHandler.java:
```diff
@@ -618,276 +606,692 @@ private void onDisconnected() {
-    @Override
-    public void updateAvaliabilityStatus(IZoneMinderConnectionInfo connection) {
+    private int initRetriesCount = 0;
+    private static int initMaxRevoverableRetries = 5;
```
Why static, is it also final? What does it mean if two processes would change it at the same time.

martinvw commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/handler/ZoneMinderServerBridgeHandler.java:
```diff
@@ -618,276 +606,692 @@ private void onDisconnected() {
+                updateStatus(newStatus, statusDetail, statusDescription);
+
+                logger.error("{}: context='{}' state='{}' check='FAILED' - {}", getLogIdentifier(), context,
+                        newStatus.toString(), statusDescription);
```
You don't have to call `toString` it will be called automatically, but only when the logging-level actually applies.

martinvw commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/handler/ZoneMinderServerBridgeHandler.java:
```diff
@@ -591,23 +578,24 @@ public void onConnected() {
-                    logger.error("{}: onConnected() failed - Exceprion: {}", getLogIdentifier(), e.getMessage());
+                    thingHandler.onBridgeConnected(this, getZoneMinderConnection());
+                } catch (IllegalArgumentException e) {
+                    logger.error("{}: context='onConnected' failed - Exceprion: {}", getLogIdentifier(),
```
Typo 'Exceprion'

martinvw commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/handler/ZoneMinderServerBridgeHandler.java:
```diff
@@ -394,111 +365,140 @@ public void handleCommand(ChannelUID channelUID, Command command) {
             } catch (Exception ex) {
-                logger.error("{}: Method 'refreshThing()' for Bridge {} failed for thing='{}' - Exception='{}'",
-                        getLogIdentifier(), this.getZoneMinderId(), thing.getUID(), ex.getMessage());
+                logger.error("{}: context='refreshThing' tag='exception' Exception thrown when refreshing thing='{}'",
```
In [E.4 of the coding guidelines](http://docs.openhab.org/developers/development/guidelines.html#e-logging) we listed what kind of logging level should be used when. Error should only be used:

> `error` logging should only be used to inform the user that something is tremendously wrong in his setup, the system cannot function normally anymore, and there is a need for immediate action. 

Please consider to change this and all other error logs to use the correct logging levels and if possible change the ONLINE/OFFLINE status of thing / bridge.

martinvw commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/handler/ZoneMinderServerBridgeHandler.java:
```diff
@@ -618,276 +606,692 @@ private void onDisconnected() {
+                logger.error("{}: context='{}' state='{}' check='FAILED' - {}", getLogIdentifier(), context,
+                        newStatus.toString(), statusDescription);
+                return status;
+            }
```
There should be no newlines between a closing curly brace and the `else`, please also review the report of [the Static code analysis](https://docs.openhab.org/developers/development/bindings.html#static-code-analysis), this also saves me some time while reviewing. I expected that there are some more references to undesired newlines and such

martinvw commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/handler/ZoneMinderServerBridgeHandler.java:
```diff
@@ -618,276 +606,692 @@ private void onDisconnected() {
+                updateStatus(newStatus, statusDetail, statusDescription);
+
+                logger.error("{}: context='{}' state='{}' check='FAILED' - {}", getLogIdentifier(), context,
+                        newStatus.toString(), statusDescription);
```
You don't have to call `toString` it will be called automatically, but only when the logging-level actually applies.

Please apply everywhere.

Note that this does not have to be error logging anymore because you already configured the thing status, it more likely that he will see that than that he will the error log.

martinvw commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/handler/ZoneMinderServerBridgeHandler.java:
```diff
@@ -618,276 +606,692 @@ private void onDisconnected() {
+                return status;
+
+            } else if (config.getHost() == null) {
+                newStatus = ThingStatus.OFFLINE;
```
This method looks to have a lot of repetition, can you implement this in smarter way?

Mr-Eskildsen replied:
The "problem" is that I need to check every parameter, to be able to log a precise error. 
All the repetetive assignments removed.

martinvw commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/handler/ZoneMinderServerBridgeHandler.java:
```diff
@@ -618,276 +606,692 @@ private void onDisconnected() {
 
-                    return;
-                }
+            newStatus = ThingStatus.OFFLINE;
```
Why OFFLINE, you are even connecting, maybe UNKNOW would be more suitable?

Mr-Eskildsen replied:
Nope, at that specific place I have verified that config in openHAB looks good, Now I have to verify the remote configuration, which could also be wrong. The only way to find out is trial and error . I will add comments in the code to explain :-)

martinvw commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/handler/ZoneMinderServerBridgeHandler.java:
```diff
@@ -618,276 +606,692 @@ private void onDisconnected() {
+                forcedPriority = RefreshPriority.UNKNOWN;
+                if (updateConnection) {
+                    try {
+                        setConnected(false);
```
If this method is not a setter, which is suggested by the number of exceptions it throws, it might be better to call it `disconnect`. Because now it confuse me as the reader

martinvw commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/state/MonitorThingState.java:
```diff
@@ -0,0 +1,537 @@
+    private boolean isDirty;
+    private String logIdentifier = "";
+
+    private IZoneMinderEventData activeEvent = null;
```
These values would also be assigned here, if you left out the `null`

Mr-Eskildsen commented on addons/binding/org.openhab.binding.zoneminder/ESH-INF/binding/binding.xml:
```diff
@@ -1,11 +1,9 @@
-	<description>This binding interfaces a ZoneMinder Server</description>
-	<author>Martin S. Eskildsen</author>
-
+        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
```
I have struggled (with no luck regarding xml files) to remove this from the analysis report. It seems that eclipse just converts to spaces when saving. Maybe I am doing something wrong?

Mr-Eskildsen commented on addons/binding/org.openhab.binding.zoneminder/README.md:
````diff
@@ -49,15 +49,48 @@ When a new monitor is discovered it will appear in the Inbox.
+### Things file
 ```
-Bridge zoneminder:server:ZoneMinderSample [ hostname="192.168.1.55", user="<USERNAME>", password="<PASSWORD>", telnet_port=6802, refresh_interval_disk_usage=1 ]
+Bridge zoneminder:server:ZoneMinderSample [ host="192.168.1.55", protocol="http", user="<USERNAME>", password="<PASSWORD>", autodiscover=false, useSpecificUserStreaming=true, streamingUser="<STREAMING-USER>", streamingPassword="<STREAMING-PASSWORD>" ]
````
You have a good point, but acutally I believe not doing it, since I have released this to the marketplace longtime ago . And I believe that most of the users is running on the marketplace version (since it is far more stable).

Mr-Eskildsen commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/handler/ZoneMinderBaseThingHandler.java:
```diff
@@ -100,123 +93,196 @@ public ZoneMinderBaseThingHandler(Thing thing) {
 
     protected void releaseSession() {
-        lockSession.unlock();
+        // lockSession.unlock();
```
Left over from the old http client, I will remove it as well as the aquireSession (which is simply just a getSession method

Mr-Eskildsen commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/handler/ZoneMinderBaseThingHandler.java:
```diff
@@ -100,123 +93,196 @@ public ZoneMinderBaseThingHandler(Thing thing) {
+    }
+
+    private IZoneMinderConnectionHandler aquireSessionInternal(boolean timeout) {
+        boolean result = true;
```
Correct - As mentioned earlier it is a leftover, while be simplified :-).

wborn commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/state/ChannelRawType.java:
```diff
@@ -0,0 +1,53 @@
+    }
+
+    @Override
+    protected State convert(Object _state) throws UnsupportedDataTypeException {
```
Underscores should not be part of variable names so rename _state.

wborn commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/state/ChannelState.java:
```diff
@@ -0,0 +1,37 @@
+    public State getState();
+
+    // private State statePublished = UnDefType.NULL;
+    // private Type dataType = UnDefType.class;
```
Enable or remove commented code.

wborn commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/state/ChannelStateChangePublisher.java:
```diff
@@ -0,0 +1,46 @@
+     *
+     * public void subscribeNumberType(ChannelUID channelUID, ChannelStateChangeSubscriber subscriber);
+     */
+    // public void subscribeRawType(ChannelUID channelUID, ChannelStateChangeSubscriber subscriber);
```
Enable or remove commented code.

wborn commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/state/ChannelStateChangePublisher.java:
```diff
@@ -0,0 +1,46 @@
+
+    public void addChannel(ChannelUID channelUID/* , ChannelStateChangeSubscriber subscriber */);
+
+    public void subscribe(ChannelUID channelUID/* , ChannelStateChangeSubscriber subscriber */);
```
Enable or remove commented code.

wborn commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/state/ChannelStateChangePublisher.java:
```diff
@@ -0,0 +1,46 @@
+                                   * ChannelStateChangeSubscriber subscriber
+                                   */);
+
+    public void addChannel(ChannelUID channelUID/* , ChannelStateChangeSubscriber subscriber */);
```
Enable or remove commented code.

wborn commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/state/ChannelStateChangePublisher.java:
```diff
@@ -0,0 +1,46 @@
+    public GenericChannelState createChannelSubscription(
+            ChannelUID channelUID /*
+                                   * ,
+                                   * ChannelStateChangeSubscriber subscriber
```
Enable or remove commented code.

wborn commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/state/ChannelStringType.java:
```diff
@@ -0,0 +1,48 @@
+    }
+
+    @Override
+    protected State convert(Object _state) throws UnsupportedDataTypeException {
```
Underscores should not be part of variable names so rename _state.

wborn commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/state/GenericThingState.java:
```diff
@@ -0,0 +1,182 @@
+    private ChannelStateChangeSubscriber subscriber = null;
+
+    // Keeps track of current channel refresh status of subscribed channels
+    private Map<String, GenericChannelState> subscriptions = new ConcurrentHashMap<String, GenericChannelState>();
```
Use the diamond operator: `new ConcurrentHashMap<>();`

wborn commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/state/GenericThingState.java:
```diff
@@ -0,0 +1,182 @@
+
+    final AtomicBoolean allowRefresh = new AtomicBoolean(true);
+
+    private ChannelStateChangeSubscriber subscriber = null;
```
Remove unnecessary null initialization.

wborn commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/state/GenericThingState.java:
```diff
@@ -0,0 +1,182 @@
+            }
+        } catch (
+
+        Exception ex) {
```
Fix indentation and please catch specific exceptions instead of catching all exceptions. That way it is more clear what exceptions need to be handled and how this is done.

wborn commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/state/GenericThingState.java:
```diff
@@ -0,0 +1,182 @@
+    }
+
+    @Override
+    public void subscribe(ChannelUID channelUID/* , ChannelStateChangeSubscriber subscriber */) {
```
Enable or remove commented code.

wborn commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/state/GenericThingState.java:
```diff
@@ -0,0 +1,182 @@
+    public void subscribe(ChannelUID channelUID/* , ChannelStateChangeSubscriber subscriber */) {
+        try {
+            if (getChannelStateHandler(channelUID) == null) {
+                addChannel(channelUID/* , subscriber */);
```
Enable or remove commented code.

wborn commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/state/GenericThingState.java:
```diff
@@ -0,0 +1,182 @@
+
+        } catch (
+
+        Exception ex) {
```
Fix indentation and please catch specific exceptions instead of catching all exceptions. That way it is more clear what exceptions need to be handled and how this is done.

wborn commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/state/GenericThingState.java:
```diff
@@ -0,0 +1,182 @@
+        return allowRefresh.get();
+    }
+
+    public void forceUpdate(boolean _recalculate) {
```
Remove underscore from variable name.

wborn commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/state/GenericThingState.java:
```diff
@@ -0,0 +1,182 @@
+
+    /**
+     *
+     * OLD STUFF Needs cleaning
```
Please clean the old stuff. :smile:

wborn commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/state/GenericThingState.java:
```diff
@@ -0,0 +1,182 @@
+    protected State getBooleanAsOnOffState(boolean value) {
+        State state = UnDefType.UNDEF;
+        state = value ? OnOffType.ON : OnOffType.OFF;
+        return state;
```
If the old stuff remains also simplify it, e.g.:

```java
return  UnDefType.UNDEF ? OnOffType.ON : OnOffType.OFF;
```

wborn commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/state/GenericThingState.java:
```diff
@@ -0,0 +1,182 @@
+            state = new StringType(value);
+        }
+
+        return state;
```
If the old stuff remains also simplify it, e.g.:

```java
return  value != null ? new StringType(value) : UnDefType.UNDEF;
```

wborn commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/state/GenericThingState.java:
```diff
@@ -0,0 +1,182 @@
+            state = new RawType(baos.toByteArray(), "image/jpeg");
+        }
+
+        return state;
```
If the old stuff remains also simplify it, e.g.:

```java
return baos != null ? new RawType(baos.toByteArray(), "image/jpeg") : UnDefType.UNDEF;
```

wborn commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/state/MonitorThingState.java:
```diff
@@ -0,0 +1,537 @@
+     * public MonitorThingState(String id, ChannelStateChangeSubscriber subscriber) {
+     * super(subscriber);
+     * logIdentifier = id;
+     * initialize();
```
Enable or remove commented code.

wborn commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/state/MonitorThingState.java:
```diff
@@ -0,0 +1,537 @@
+                    break;
+            }
+
+        } catch (Exception ex) {
```
Please catch specific exceptions instead of catching all exceptions. That way it is more clear what exceptions need to be handled and how this is done.

wborn commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/state/MonitorThingState.java:
```diff
@@ -0,0 +1,537 @@
+        if (stateEnabled == OnOffType.ON) {
+            return true;
+        }
+        return false;
```
Simplify to:

```java
return stateEnabled == OnOffType.ON;
```

wborn commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/state/MonitorThingState.java:
```diff
@@ -0,0 +1,537 @@
+
+    public boolean getMonitorEventMotion() {
+        State stateEventMotion = getChannelStateHandler(ZoneMinderConstants.CHANNEL_MONITOR_MOTION_EVENT).getState();
+        return false;
```
Looks like a bug because the `stateEventMotion` variable is not used. :roll_eyes:

wborn commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/state/MonitorThingState.java:
```diff
@@ -0,0 +1,537 @@
+
+        try {
+            result = ZoneMinderMonitorStatusEnum.getEnumFromName(stateStatus.toString());
+        } catch (Exception ex) {
```
Please catch specific exceptions instead of catching all exceptions. That way it is more clear what exceptions need to be handled and how this is done.

wborn commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/state/MonitorThingState.java:
```diff
@@ -0,0 +1,537 @@
+            logger.error("{}: context='setMonitorEnabled' Exception occurred when calling setMonitorEnabled",
+                    logIdentifier, e);
+
+        } catch (Exception ex) {
```
Please catch specific exceptions instead of catching all exceptions. That way it is more clear what exceptions need to be handled and how this is done.

wborn commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/state/MonitorThingState.java:
```diff
@@ -0,0 +1,537 @@
+                    setMonitorEventMotion(OnOffType.OFF);
+                    break;
+            }
+        } catch (Exception ex) {
```
Please catch specific exceptions instead of catching all exceptions. That way it is more clear what exceptions need to be handled and how this is done.

wborn commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/state/MonitorThingState.java:
```diff
@@ -0,0 +1,537 @@
+                state = new RawType(image.toByteArray(), "image/jpeg");
+            }
+
+        } catch (Exception ex) {
```
Please catch specific exceptions instead of catching all exceptions. That way it is more clear what exceptions need to be handled and how this is done.

wborn commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/state/MonitorThingState.java:
```diff
@@ -0,0 +1,537 @@
+                break;
+        }
+    }
+}
```
Each file should end with a new line.

wborn commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/handler/ZoneMinderBaseThingHandler.java:
```diff
@@ -367,7 +420,8 @@ protected State getChannelStringAsStringState(String channelValue) {
             }
 
         } catch (Exception ex) {
```
Please catch specific exceptions instead of catching all exceptions. That way it is more clear what exceptions need to be handled and how this is done.

wborn commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/handler/ZoneMinderBaseThingHandler.java:
```diff
@@ -383,25 +437,28 @@ protected State getChannelBoolAsOnOffState(boolean value) {
             }
 
         } catch (Exception ex) {
```
Please catch specific exceptions instead of catching all exceptions. That way it is more clear what exceptions need to be handled and how this is done.

wborn commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/handler/ZoneMinderServerBridgeHandler.java:
```diff
@@ -123,211 +144,173 @@ public void run() {
-                    logger.error("[MONITOR]: Method 'refreshThing()' for Bridge failed for thing='{}' - Exception: ",
-                            thing.getUID(), ex);
-                }
+            } catch (Exception exception) {
```
Please catch specific exceptions instead of catching all exceptions. That way it is more clear what exceptions need to be handled and how this is done.

wborn commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/handler/ZoneMinderServerBridgeHandler.java:
```diff
@@ -394,111 +365,140 @@ public void handleCommand(ChannelUID channelUID, Command command) {
+                            logger.debug("{}: URL='{}' ResponseCode='{}' ResponseMessage='{}'", getLogIdentifier(),
+                                    diskUsage.getHttpRequestUrl(), diskUsage.getHttpStatus(),
+                                    diskUsage.getHttpResponseMessage());
+                        } catch (Exception ex) {
```
Please catch specific exceptions instead of catching all exceptions. That way it is more clear what exceptions need to be handled and how this is done.

wborn commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/handler/ZoneMinderServerBridgeHandler.java:
```diff
@@ -394,111 +365,140 @@ public void handleCommand(ChannelUID channelUID, Command command) {
+                channelCpuLoad = "";
+                channelDiskUsage = "";
+            }
+        } catch (Exception ex) {
```
Please catch specific exceptions instead of catching all exceptions. That way it is more clear what exceptions need to be handled and how this is done.

wborn commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/handler/ZoneMinderServerBridgeHandler.java:
```diff
@@ -511,63 +511,50 @@ public synchronized Boolean isConnected() {
+                        }
 
-    }
+                    } catch (Exception ex) {
```
Please catch specific exceptions instead of catching all exceptions. That way it is more clear what exceptions need to be handled and how this is done.

wborn commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/handler/ZoneMinderServerBridgeHandler.java:
```diff
@@ -618,276 +606,692 @@ private void onDisconnected() {
+                    if (thingHandler instanceof ZoneMinderThingMonitorHandler) {
+                        try {
+                            thingHandler.updateAvaliabilityStatus(getZoneMinderConnection());
+                        } catch (Exception ex) {
```
Please catch specific exceptions instead of catching all exceptions. That way it is more clear what exceptions need to be handled and how this is done.

wborn commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/handler/ZoneMinderServerBridgeHandler.java:
```diff
@@ -901,58 +1305,91 @@ public void updateChannel(ChannelUID channel) {
-                        getLogIdentifier(), channel.getId(), state.toString());
                 updateState(channel.getId(), state);
             }
         } catch (Exception ex) {
```
Please catch specific exceptions instead of catching all exceptions. That way it is more clear what exceptions need to be handled and how this is done.

wborn commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/handler/ZoneMinderThingMonitorHandler.java:
```diff
@@ -117,63 +118,119 @@ public String getZoneMinderId() {
+
+            logger.info("{}: context='channelLinked' - Unlinking from channel '{}'", getLogIdentifier(),
+                    channelUID.getAsString());
         } catch (Exception ex) {
```
Please catch specific exceptions instead of catching all exceptions. That way it is more clear what exceptions need to be handled and how this is done.

wborn commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/handler/ZoneMinderThingMonitorHandler.java:
```diff
@@ -185,112 +242,193 @@ public void handleCommand(ChannelUID channelUID, Command command) {
+                                    }
+                                    fetchMonitorGeneralData(monitorProxy);
+
+                                } catch (Exception ex) {
```
Please catch specific exceptions instead of catching all exceptions. That way it is more clear what exceptions need to be handled and how this is done.

wborn commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/handler/ZoneMinderThingMonitorHandler.java:
```diff
@@ -310,19 +448,28 @@ else if (command == OnOffType.OFF) {
+            dataConverter.addChannel(getChannelUIDFromChannelId(ZoneMinderConstants.CHANNEL_MONITOR_FUNCTION));
+            dataConverter.addChannel(getChannelUIDFromChannelId(ZoneMinderConstants.CHANNEL_MONITOR_EVENT_STATE));
+
         } catch (Exception ex) {
```
Please catch specific exceptions instead of catching all exceptions. That way it is more clear what exceptions need to be handled and how this is done.

wborn commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/handler/ZoneMinderThingMonitorHandler.java:
```diff
@@ -332,24 +479,73 @@ public void initialize() {
+                            eventData.getHttpRequestUrl(), eventData.getHttpStatus(),
+                            eventData.getHttpResponseMessage());
+
+                } catch (Exception ex) {
```
Please catch specific exceptions instead of catching all exceptions. That way it is more clear what exceptions need to be handled and how this is done.

wborn commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/handler/ZoneMinderThingMonitorHandler.java:
```diff
@@ -624,265 +739,451 @@ public void updateStatus(ThingStatus status) {
-            logger.debug("{}: URL='{}' ResponseCode='{}' ResponseMessage='{}'", getLogIdentifier(),
-                    monitorProxy.getHttpUrl(), monitorProxy.getHttpResponseCode(),
-                    monitorProxy.getHttpResponseMessage());
+        } catch (Exception ex) {
```
Please catch specific exceptions instead of catching all exceptions. That way it is more clear what exceptions need to be handled and how this is done.

wborn commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/handler/ZoneMinderThingMonitorHandler.java:
```diff
@@ -624,265 +739,451 @@ public void updateStatus(ThingStatus status) {
+                    logger.error(
+                            "{}: context='fetchMonitorDaemonStatus' error in call to 'getAnalysisDaemonStatus' - Exception='{}' ",
+                            getLogIdentifier(), zme.getClass().getCanonicalName(), zme.getCause());
+                } catch (Exception ex) {
```
Please catch specific exceptions instead of catching all exceptions. That way it is more clear what exceptions need to be handled and how this is done.

wborn commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/handler/ZoneMinderThingMonitorHandler.java:
```diff
@@ -624,265 +739,451 @@ public void updateStatus(ThingStatus status) {
+                    logger.error(
+                            "{}: context='fetchMonitorDaemonStatus' error in call to 'getFrameDaemonStatus' - Exception='{}'",
+                            getLogIdentifier(), zme.getClass().getCanonicalName(), zme.getCause());
+                } catch (Exception ex) {
```
Please catch specific exceptions instead of catching all exceptions. That way it is more clear what exceptions need to be handled and how this is done.

wborn commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/handler/ZoneMinderThingMonitorHandler.java:
```diff
@@ -893,14 +1194,28 @@ public String getLogIdentifier() {
                 result = String.format("[MONITOR-%s]", config.getZoneMinderId().toString());
             }
 
         } catch (Exception ex) {
```
Please catch specific exceptions instead of catching all exceptions. That way it is more clear what exceptions need to be handled and how this is done.

wborn commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/state/GenericThingState.java:
```diff
@@ -0,0 +1,182 @@
+            if (getChannelStateHandler(channelUID) != null) {
+                getChannelStateHandler(channelUID).subscribe();
+            }
+        } catch (Exception ex) {
```
Please catch specific exceptions instead of catching all exceptions. That way it is more clear what exceptions need to be handled and how this is done.

wborn commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/handler/ZoneMinderThingMonitorHandler.java:
```diff
@@ -310,19 +448,28 @@ else if (command == OnOffType.OFF) {
-            logger.debug("{}:    Monitor Id:         {}", getLogIdentifier(), config.getZoneMinderId());
+
+            super.initialize();
+            logger.info("{}: context='initialize' Monitor Handler Initialized", getLogIdentifier());
```
Reduce to debug.

wborn commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/state/ChannelStateChangeSubscriber.java:
```diff
@@ -0,0 +1,26 @@
+import org.eclipse.smarthome.core.types.State;
+
+/**
+ * The {@link GenericThingState} is responsible for handling commands, which are
```
GenericThingState? Please add an appropriate JavaDoc.

wborn commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/state/ChannelStringType.java:
```diff
@@ -0,0 +1,48 @@
+import org.eclipse.smarthome.core.types.UnDefType;
+
+/**
+ * The {@link GenericThingState} is responsible for handling commands, which are
```
GenericThingState? Please add an appropriate JavaDoc.

wborn commented on addons/binding/org.openhab.binding.zoneminder/src/main/java/org/openhab/binding/zoneminder/internal/state/GenericChannelState.java:
```diff
@@ -0,0 +1,111 @@
+import org.eclipse.smarthome.core.types.UnDefType;
+
+/**
+ * The {@link GenericThingState} is responsible for handling commands, which are
```
GenericThingState? Please add an appropriate JavaDoc.

martinvw commented:
@Mr-Eskildsen thanks for your contribution, do you expect to be able to wrap up this PR? Thanks for keeping us updated!

wborn commented:
Any plans for addressing the review comments @Mr-Eskildsen ? We've got several PRs waiting on this PR. Or is it OK with you that we merge #3382 and #4037?

Mr-Eskildsen commented:
Sorry, I am little bit stucked with this. I think we should just reject the
PR.
/Martin

Den fre. 25. jan. 2019 kl. 21.55 skrev Wouter Born <notifications@github.com
>:

> Any plans for addressing the review comments @Mr-Eskildsen
> <https://github.com/Mr-Eskildsen> ? We've got several PRs waiting on this
> PR. Or is it OK with you that we merge #3382
> <https://github.com/openhab/openhab2-addons/pull/3382> and #4037
> <https://github.com/openhab/openhab2-addons/pull/4037>?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openhab/openhab2-addons/pull/3583#issuecomment-457724707>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFTMexVn9jJ8iQ2oE6ViNufRCaiDWUBNks5vG28zgaJpZM4UG8b2>
> .
>

wborn commented:
Thanks for the update! :+1:  We'll then merge the other two PRs. It would be awesome if you can still find some time at a later moment to continue your great work on this PR!

J-N-K commented:
Any progress here? We would like to move the binding to the bnd build system.

davidgraeff commented:
I'd say we close this for now. @Mr-Eskildsen maybe you can copy paste part of the this PR back to master later on and provide smaller pull requests that result in faster merges :)
